admin_only: default to full_admin semantics

joshmoore opened the pull request:
The semantics of the admin_only have been changed by the
introduction of restricted admins. Whereas previously,
isAdmin (i.e. membership in the system group) implied
complete power, restricted admins are *also* in the
system group but may be missing a non-future-proof list
of privileges.

This restores the previous behavior of having admin_only
require full permissions by acquiring the current list
of privileges and checking against those of the user.

If the new "minimial-restricted-admin" logic is needed,
then use `@admin_only(full_admin=False)`.

See:
 * https://trello.com/c/4dvUYSjS/126-definition-of-ultimate-admin
 * https://github.com/openmicroscopy/openmicroscopy/pull/5494

joshmoore committed:
admin_only: default to full_admin semantics

The semantics of the admin_only have been changed by the
introduction of restricted admins. Whereas previously,
isAdmin (i.e. membership in the system group) implied
complete power, restricted admins are *also* in the
system group but may be missing a non-future-proof list
of privileges.

This restores the previous behavior of having admin_only
require full permissions by acquiring the current list
of privileges and checking against those of the user.

If the new "minimial-restricted-admin" logic is needed,
then use `@admin_only(full_admin=False)`.

See: https://trello.com/c/4dvUYSjS/126-definition-of-ultimate-admin

mtbc commented:
Does this leave us any way to say, "must be an admin, but no particular restrictions are relevant"? *Update:* Is this is the "minimial-restricted-admin" logic?

mtbc commented:
Affected are:

* admin fixpyramids
* fs set_repo (which is nothing?)
* ldap active, list, getdn, setdn, discover

Probably it is manageable to review these cases to figure what we want to do. Especially, for which are we comfortable with `full_admin=False`? (Probably pretty much all?) (`setdn` has extra logic depending on if user or group is targeted.)

mtbc commented:
Related followup question that need not block 5.4.0: What would it take to relax those remaining `@admin_only()` in a future-proof way? Would a `boolean IAdmin.canReadAllData()` help?

pwalczysko commented:
Fully agree with @mtbc 's [suggestion](https://github.com/will-moore/openmicroscopy/pull/58#issuecomment-327782614). Lets put the ``full_admin=False`` on these commands and go ahead with the change.

joshmoore committed:
Review all `admin_only()` usages

This opens up the access to some of the admin_only decorators
by passing `full_admin=False`.

joshmoore commented:
Some slight disagreement. Pushed these updates:

 * `admin fixpyramids`: I disagree. This is accessing the repo and should be protected
 * `fs set_repo`: this does nothing and I've removed it.
 * `ldap active`:  this for me points to a future `Config` privilege.
 * `ldap list`: I don't know why this is admin_only. The API methods are `RolesAllowed("user")`. I've removed but I could also make the server method more restrictive.
 * `ldap getdn`: this uses `ILdap` which is RolesAllowed("system")
 * `ldap setdn`: Making this permissive since the lower-level methods are less tolerant.
 * `discover`: again, this falls under `ILdap`'s behavior which isn't well defined at the moment.

mtbc commented:
Thank you for this review.

pwalczysko commented:
@joshmoore : is it possible that this PR broke https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-integration-Python27/lastCompletedBuild/testReport/OmeroPy.test.integration.clitest.test_ldap/TestLDAP/testAdminOnly_list_/ ?
